### ISSUE TITLE
Adding getServiceLocator() to ServiceLocatorInterface

### ISSUE THREAD

**Issue by manuakasam** [zendframework/zf2#7388]
Hey guys,

so I was just reading the newest Blog by @samsonasik and it got me thinking. Here a link to the article https://samsonasik.wordpress.com/2015/03/31/zend-framework-2-using-__invokepluginmanager-manager-in-services-factory/

Basically it teaches the approach of using Invokable classes as ServiceFactories. And I don't like it.

# Where is the Problem?

As a community we are teaching newcomers to differentiate between two kinds of services: `invokables` and `factories`. If now the community starts to only use `invokables` because of proper type-hinting, then newcomers will simply be blown away with a bit of "wtf?".

While it is perfectly fine to come up with new ways of doing things, I feel that the problem is of a different kind of nature and that we should rather focus on solving said problem than to find workarounds for it.

# Analyzing the Problem

Service-Factories for the default `ServiceManager` we're writing a Factory like this:

```php
class FooFactory implements FactoryInterface
{
    public function createService(ServiceLocatorInterface $sl) 
    {
        return new FooService(
            $sl->get('FooDependency')
        );
    }
}
```

All nice and clean. Both parties, Users and Code-Inspection, are very happy.

However when we're outside the scope of the normal `ServiceManager`, for example when writing Controller-Factories, things are changing. Take a look at this example:

```php
class BarControllerFactory implements FactoryInterface
{
    public function createService(ServiceLocatorInterface $sl) 
    {
        /** @var $realSl ControllerManager
        $realSl = $sl->getServiceLocator();
        
        return new BarController(
            $realSl->get('BarDependency')
        );
    }
}
```

And to me this is where we run into issues. While we're successfully tricking our IDE into giving us proper code-completion, code-inspection tools like Scrutinizer are analyzing the variable based on what we pass into the factory. That that's only the `ServiceLocatorInterface`. An interface that does not know about `getServiceLocator()`.

From a programming point of view I do understand why we're using the shared interface. However the issue that I have - and I'm sure I can't be the only one - is, that we're relying on something that isn't defined by the interface. In essence that's bad design.

# What can be done?

Now I haven't given this many thorough thoughts yet, but as far as a quick brainstorm is concerned I see two possible solutions.

## Create new Interfaces

This would mean we have a new Interface for each of the `ServiceManager`-aware Managers like the `ControllerManager`, `FormManager`, etc.., etc... This interface would then implement `getServiceLocator()`. In turn this would allow for  a better code-only-documentation and both users and code-tools would be more happy.
 
## Add getSl to the ServiceLocatorInterface

This change - in my opinion - would be the one that has the least of effects. When adding `getServiceLocator()` to the `ServiceLocatorInterface` the problem with the hidden dependency on all other factories would be solved instantly. The new issue it raises is that the `ServiceManager` itself also implements the `ServiceLocatorInterface`. So we'd have to implement it like this:

```php
class ServiceManager implements ServiceLocatorInterface
{
    public function getServiceLocator() 
    {
        return $this;
    }
    
    // rest of class...
}
```

This **obviously** isn't ideal. It would allow for ridiculous code-chaining like: 

```
$sl->getServiceLocator()->getServiceLocator()->getServiceLocator()->getServiceLocator()->get('RickRolled')
```

though personally I don't see it of this much as an issue. Would anyone really do this?

# Summary

To sum it up:

- The Framework implies the usage of something that isn't defined by an interface
- Workarounds do exist but the raise yet another new learning barrier for starters
- In my opinion we should change this fact, the way to do so should be open for discussion

What are your thoughts about this? Do you share the concern? Got alternate solutions? Is this an issue at all?

Feedback - as always - much appreciated.

**Comment by Martin-P** [comment]
> As a community we are teaching newcomers to differentiate between two kinds of services: `invokables` and `factories`.

Most people don't even know about the feature with invokable classes (https://github.com/zendframework/zf2/pull/7133#issuecomment-70425781). When you are lucky people use factories, but because the manual uses closures people tend to use closures :cry: 

- [Using ServiceManager to configure the table gateway and inject into the AlbumTable](http://framework.zend.com/manual/current/en/user-guide/database-and-models.html#using-servicemanager-to-configure-the-table-gateway-and-inject-into-the-albumtable)
- [Using Service Manager to configure the database credentials and inject into the controller](http://framework.zend.com/manual/current/en/getting-started-with-zend-studio/the-application.html#using-service-manager-to-configure-the-database-credentials-and-inject-into-the-controller)

In my own code I keep it clean by using an other name when I expect a `ControllerManager`:
```php
class BarControllerFactory implements FactoryInterface
{
    public function createService(ServiceLocatorInterface $controllerManager) 
    {
        $serviceManager = $controllerManager->getServiceLocator();
        // etc.
    }
}
```

> This would mean we have a new Interface for each of the `ServiceManager`-aware Managers like the `ControllerManager`, `FormManager`, etc.., etc...

You would need separate interfaces like a `ControllerManagerFactoryInterface` or `FormManagerFactoryInterface`. You can use autocompletion in your IDE that way, but when you are aware of the "trick" that you get a `ControllerManager` in a factory for a controller, a `ViewManager` in a factory for a view helper, etc. I don't know if the new interfaces make it easier or just harder.

**Comment by samsonasik** [comment]
using <code>ControllerManagerFactoryInterface</code> won't work as <code>createService</code> requires <code>ServiceLocatorInterface</code>, imo, we can use new factory interface for managers that passed AbstractPluginManager.
```php
interface FactoryManagerInterface
{
    public function createService(AbstractPluginManager $serviceLocator);
}

class BarControllerFactory implements FactoryManagerInterface
{
    public function createService(AbstractPluginManager $serviceLocator)
    {
        // ..
    }
}
```

**Comment by manuakasam** [comment]
@samsonasik while this is one of the approaches I outlined, I'm not truly happy with it. While it eases the learning curve for beginners on all other PluginManagers, now we do have two factoryInterfaces and we need to teach that for ServiceManager-Factories you need to use Interface A and for PluginManager-Factories you need to use Interface B. While likely the most clean approach from a programming perspective, the usability is going down. With that said, I'd still be happier with your suggestion than I am with the current implementation :)

**Comment by Ocramius** [comment]
Note: the entire `getServiceLocator` thing is totally nuts. I actually would prefer to categorize services inside a single locator, if that makes sense.

**Comment by manuakasam** [comment]
@Ocramius with that being said, do you recall the reasons to separate the SL into multiple DI containers?

**Comment by Martin-P** [comment]
> imo, we can use new factory interface for managers that passed AbstractPluginManager

Which brings us back to the same issue, except now it is for `AbstractPluginManager` instead of `ServiceLocatorInterface` :wink:

**Comment by Ocramius** [comment]
The problem being solved was very simple: filtering available services.

What we had (when releasing `2.0.0-beta2`) was a security issue with routes like `/foo/:controller`.

This sort of route is problematic, as it is possible to request any controller via the URI string, which means that also (potentially) malicious classes may be instantiated/initialized.

Actions that were taken:
 - segregating controllers into their own `ControllerManager` (an `AbstractPluginManager`)
 - disable `autoInvokables`

That's why we have plugin managers in first place.

The concept was extended to allow simple type strictness on services, but it doesn't really work as expected, and it requires a better solution, IMO.

One way to solve this issue, in my opinion, would be:

 - tagged services (much like symfony does it)
 - introducing a per-tag service validator (which may act pre- or post-instantiation)

**Comment by bakura10** [comment]
In my refactor (#7133) I took a different approach to this problem to remove the crap of `getServiceLocator`.

I realized that in most of my factories, a plugin manager factory needs dependencies outside of the plugin manager (therefore, dependencies coming from main service locator, or other plugin managers). It's very rare that I need other dependencies coming from the same manager.

Therefore in the refactor I removed this `getServiceLocator`, and the main service locator MUST be set as part of the plugin manager (https://github.com/bakura10/zf2/blob/zf3-sm/library/Zend/ServiceMananger/AbstractPluginManager.php#L30). In ZF2, you had to set it through a setter, which many people forgot to do, hence causing hard to find bugs.

The interesting thing is that this is passed as a so-called "creationContext" (https://github.com/bakura10/zf2/blob/zf3-sm/library/Zend/ServiceMananger/AbstractPluginManager.php#L33).

This creation context is passed as the first parameter of each factory: https://github.com/bakura10/zf2/blob/zf3-sm/library/Zend/ServiceMananger/ServiceManager.php#L106

No need for calling `getServiceLocator`, as you'll always receive the main service locator as first parameter.

**Comment by bakura10** [comment]
I'm not against removing plugin manager completely as @Ocramius suggested though. Although I like them as injecting a plugin manager as a dependency is kinda accepted, while injecting the whole service locator is evil (although, at the end, it's quite similar as the service locator can still be accessed).

**Comment by manuakasam** [comment]
I like what you did @bakura10 and I totally agree with your impression "factories rarely require dependencies from the same plugin manager but rather from the main service locator". 

So what's the current "chance of success" for your rewrite being the ZF2.5/3 thingy? Because if those are quite high then I would totally see this issue here as a non-issue anymore and close it.

**Comment by bakura10** [comment]
It seems the debate is now for removing plugin manager or keeping them. +1 from me for keeping them. And also have others opinion about the creationContext. It will indeed confuse existing users to see this major BC break as the service locator received will only be the main service locator, but for newcomers, it makes things much easier, as they don't have to remember always "am I in plugin manager context?".

**Comment by RalfEggert** [comment]
Should be marked for ZF3 milestone, shouldn't it?

**Comment by Maks3w** [comment]
ServiceManager V3 is merged https://github.com/zendframework/zend-servicemanager/tree/develop. Please review and if needed open an issue on the repo issue tracker.
